Updated graph api calls using Microsoft Graph SDK (#38)
Azure-Samples/active-directory-dotnet-native-uwp-v2: README.md

## About this sample

### Table of contents
- [About this sample](#about-this-sample)
  - [Overview](#overview)
- [Steps to run](#steps-to-run)
- [How to run this sample](#how-to-run-this-sample)

>   - [Step 1:  Clone or download this repository](#step-1-clone-or-download-this-repository)

Reviewer: ```suggestion
  - [Step 1: Clone or download this repository](#step-1-clone-or-download-this-repository)

```

>   - [Step 2:  Register the sample application with your Azure Active Directory tenant](#step-2-register-the-sample-application-with-your-azure-active-directory-tenant)

Reviewer: ```suggestion
  - [Step 2: Register the sample application with your Azure Active Directory tenant](#step-2-register-the-sample-application-with-your-azure-active-directory-tenant)

```

>   - [Step 4: Run the sample](#step-4-run-the-sample)

Reviewer: ```suggestion
  - [Step 3: Run the sample](#step-3-run-the-sample)

```

Author: Updated Readme to contain Step 3.
Thanks!

## About this sample

> ### Table of content

Reviewer 2: ```suggestion
### Table of contents

```

- [About this sample](#about-this-sample)

>   - [Table of content](#table-of-content)

Reviewer 2: ```suggestion
  - [Table of contents](#table-of-contents)

```

Author: Removed from TOC as it's not required.
Thanks!

  - [Overview](#overview)
- [Steps to run](#steps-to-run)
- [How to run this sample](#how-to-run-this-sample)
  - [Step 1:  Clone or download this repository](#step-1-clone-or-download-this-repository)
  - [Step 2:  Register the sample application with your Azure Active Directory tenant](#step-2-register-the-sample-application-with-your-azure-active-directory-tenant)
  - [Step 4: Run the sample](#step-4-run-the-sample)
  - [Known limitation: on Windows 10, you cannot sign in with your windows hello PIN](#known-limitation-on-windows-10-you-cannot-sign-in-with-your-windows-hello-pin)
- [Alternate approach to use WithDefaultRedirectURI()](#alternate-approach-to-use-withdefaultredirecturi)
- [Steps to build from scratch](#steps-to-build-from-scratch)
- [Community help and support](#community-help-and-support)
- [Contributing](#contributing)
- [More information](#more-information)

> 

Reviewer 2: do you need to have the part about changing the manifest? there are directions for doing some additional manual steps in the json file...should that go here as well?

Author: Manual steps are required to change the **signInAudience** and **accessTokenAcceptedVersion** in Manifest when application is registered using PowerShell scripts. 
If one follows the manual approach to register application, then the Manifest is automatically updated by selecting **Accounts in any organizational directory and personal Microsoft accounts (e.g. Skype, Xbox, Outlook.com)**.
Thanks!

### Overview

> This simple sample demonstrates how to use the [Microsoft Authentication Library (MSAL) for .NET](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet) to get an access token and call the Microsoft Graph using MS Graph SDK from an Universal Windows Platform (UWP) application.

Reviewer 2: ```suggestion
This sample demonstrates how to use the [Microsoft Authentication Library (MSAL) for .NET](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet) to get an access token and call the Microsoft Graph using the MS Graph SDK from a Universal Windows Platform (UWP) application.

```

> 1. The .NET client UWP application uses the Microsoft Authentication Library (MSAL) to sign-in a user and obtain a JWT access token from Azure Active Directory (Azure AD).

Reviewer 2: should you mention the AT is scoped...it's scoped to `user.read`?

To run this sample, you'll need:

> - [Visual Studio 2019](https://aka.ms/vsdownload)

Reviewer 3: Visual Studio with no version maybe ?

Reviewer 4: Shama, work with Tiago to update the Code generator

---
In reply to: [429118036](https://github.com/Azure-Samples/active-directory-dotnet-native-uwp-v2/pull/38#discussion_r429118036) [](ancestors = 429118036)

Reviewer 5: @jmprieur .net core 3 samples require VS2019. Should we keep the version to avoid people not being able to run the project due to the framework version?

   - If you don't have a platform added, select **Add a platform** and select the **Public client (mobile & desktop)** option.
   - In the **Redirect URIs** | **Suggested Redirect URIs for public clients (mobile, desktop)** section, select **https://login.microsoftonline.com/common/oauth2/nativeclient**

1. Select **Save** to save your changes.
1. In the app's registration screen, click on the **API permissions** blade in the left to open the page where we add access to the Apis that your application needs.
   - Click the **Add a permission** button and then,
   - Ensure that the **Microsoft APIs** tab is selected.
   - In the *Commonly used Microsoft APIs* section, click on **Microsoft Graph**
   - In the **Delegated permissions** section, select the **User.Read** in the list. Use the search box if necessary.
   - Click on the **Add permissions** button at the bottom.

##### Configure the  uwpApp app (UWP-App-calling-MSGraph) to use your app registration

Open the project in your IDE (like Visual Studio) to configure the code.
>In the steps below, "ClientID" is the same as "Application ID" or "AppId".

1. Open the `active-directory-dotnet-native-uwp-v2\MainPage.xaml.cs` file

> 1. Find the line ```private const string ClientId = "[Application Id pasted from the application registration portal]"``

Reviewer: ```suggestion
1. Find the line ```private const string ClientId = "[Application Id pasted from the application registration portal]```

```

> 1. Find the line ```private const string ClientId = "[Application Id pasted from the application registration portal]"``

Reviewer: ```suggestion
1. Find the line ```private const string ClientId = "[Application Id pasted from the application registration portal]"```

```

> ` and replace the existing value with the application ID (clientId) of the `UWP-App-calling-MSGraph` application copied from the Azure portal.

Reviewer: ```suggestion
and replace the existing value with the application ID (clientId) of the `UWP-App-calling-MSGraph` application copied from the Azure portal.

```

To run this sample, you'll need:

- [Visual Studio 2019](https://aka.ms/vsdownload)
- An Internet connection
- An Azure Active Directory (Azure AD) tenant. For more information on how to get an Azure AD tenant, see [How to get an Azure AD tenant](https://azure.microsoft.com/documentation/articles/active-directory-howto-tenant/)
- A user account in your Azure AD tenant. This sample will not work with a Microsoft account (formerly Windows Live account). Therefore, if you signed in to the [Azure portal](https://portal.azure.com) with a Microsoft account and have never created a user account in your directory before, you need to do that now.

> 

Reviewer: ```suggestion
- AzureAD Powershell Module if opting to use the automatic set up in Step 2 (available at the [Powershell Gallery](https://www.powershellgallery.com/packages/AzureAD/))

```

> ### Step 1:  Clone or download this repository

Reviewer: ```suggestion
### Step 1: Clone or download this repository

```

From your shell or command line:

```Shell
git clone https://github.com/Azure-Samples/active-directory-dotnet-native-uwp-v2.git
```

or download and extract the repository .zip file.

> Given that the names of the referenced NuGet packages are quiet long, you might want to clone it in a folder close to the root of your hard drive, to avoid file size limitations on Windows.

> ### Step 2:  Register the sample application with your Azure Active Directory tenant

Reviewer: ```suggestion
### Step 2: Register the sample application with your Azure Active Directory tenant

```

There is one project in this sample. To register it, you can:

> - either follow the steps [Step 2: Register the sample with your Azure Active Directory tenant](#step-2-register-the-sample-with-your-azure-active-directory-tenant) and [Step 3:  Configure the sample to use your Azure AD tenant](#choose-the-azure-ad-tenant-where-you-want-to-create-your-applications)

Reviewer: ```suggestion
- either follow the steps [Step 2: Register the sample with your Azure Active Directory tenant](#step-2-register-the-sample-with-your-azure-active-directory-tenant)

```

> - either follow the steps [Step 2: Register the sample with your Azure Active Directory tenant](#step-2-register-the-sample-with-your-azure-active-directory-tenant) and [Step 3:  Configure the sample to use your Azure AD tenant](#choose-the-azure-ad-tenant-where-you-want-to-create-your-applications)

Reviewer: ```suggestion
- either register the sample with your Azure AD tenant and configure it manually
```

Author: Changed Step 3 link and updated accordingly.
Thanks

- or use PowerShell scripts that:
  - **automatically** creates the Azure AD applications and related objects (passwords, permissions, dependencies) for you. Note that this works for Visual Studio only.
  - modify the Visual Studio projects' configuration files.

<details>
  <summary>Expand this section if you want to use this automation:</summary>

1. On Windows, run PowerShell and navigate to the root of the cloned directory
1. In PowerShell run:

   ```PowerShell
   Set-ExecutionPolicy -ExecutionPolicy RemoteSigned -Scope Process -Force
   ```

1. Run the script to create your Azure AD application and configure the code of the sample application accordingly.
1. In PowerShell run:

   ```PowerShell
   cd .\AppCreationScripts\
   .\Configure.ps1
   ```

   > Other ways of running the scripts are described in [App Creation Scripts](./AppCreationScripts/AppCreationScripts.md)
   > The scripts also provide a guide to automated application registration, configuration and removal which can help in your CI/CD scenarios.

1. Open the Visual Studio solution and click start to run the code.

</details>

> Follow the steps below to manually walk through the steps to register and configure the applications.

Reviewer: ```suggestion
<details>
  <summary>Expand this section to manually walk through the steps to register and configure the applications:</summary>
```

Author: As per current approach to create Readme files, this section is displayed like this.
Thanks!

Reviewer: Thanks for letting me know. Is there a guideline document you can point me to so I can read it?

Author: These steps are generated by **Code Generator**. 
Thanks!

1. Select **Save** to save your changes.
1. In the app's registration screen, click on the **API permissions** blade in the left to open the page where we add access to the APIs that your application needs.
   - Click the **Add a permission** button and then,
   - Ensure that the **Microsoft APIs** tab is selected.
   - In the *Commonly used Microsoft APIs* section, click on **Microsoft Graph**
   - In the **Delegated permissions** section, select **User.Read** in the list. Use the search box if necessary.
   - Click on the **Add permissions** button at the bottom.

##### Configure the  UWP App (UWP-App-calling-MSGraph) to use your app registration

Open the project in your IDE (like Visual Studio) to configure the code.
>In the steps below, "ClientID" is the same as "Application ID" or "AppId".

1. Open the `Native_UWP_V2\MainPage.xaml.cs` file
1. Find the line ```private const string ClientId = "[Application Id pasted from the application registration portal]"``
` and replace the existing value with the application ID (clientId) of the  `UWP-App-calling-MSGraph` application copied from the Azure portal.

> 

Reviewer: ```suggestion
</details>
```

Author: Manual setup does not contain expandable section as per current implementation.
Thanks!

To run this sample, you'll need:

- [Visual Studio 2019](https://aka.ms/vsdownload)
- An Internet connection
- An Azure Active Directory (Azure AD) tenant. For more information on how to get an Azure AD tenant, see [How to get an Azure AD tenant](https://azure.microsoft.com/documentation/articles/active-directory-howto-tenant/)
- A user account in your Azure AD tenant. This sample will not work with a Microsoft account (formerly Windows Live account). Therefore, if you signed in to the [Azure portal](https://portal.azure.com) with a Microsoft account and have never created a user account in your directory before, you need to do that now.

### Step 1:  Clone or download this repository

From your shell or command line:

```Shell
git clone https://github.com/Azure-Samples/active-directory-dotnet-native-uwp-v2.git
```

or download and extract the repository .zip file.

> > Given that the name of the sample is quiet long, and so are the names of the referenced NuGet packages, you might want to clone it in a folder close to the root of your hard drive, to avoid file size limitations on Windows.

Reviewer 2: is this still an issue now that you've shortened the name? is not, might remove the comment. idd

Author: Updated the comment to mention "names of the referenced NuGet packages" only.
Thanks!

   - Under **Supported account types**, select **Accounts in any organizational directory and personal Microsoft accounts (e.g. Skype, Xbox, Outlook.com)**.
1. Select **Register** to create the application.
1. In the app's registration screen, find and note the **Application (client) ID**. You use this value in your app's configuration file(s) later in your code.
1. In the app's registration screen, select **Authentication** in the menu.
   - If you don't have a platform added, select **Add a platform** and select the **Public client (mobile & desktop)** option.
   - In the **Redirect URIs** | **Suggested Redirect URIs for public clients (mobile, desktop)** section, select **https://login.microsoftonline.com/common/oauth2/nativeclient**

1. Select **Save** to save your changes.
1. In the app's registration screen, click on the **API permissions** blade in the left to open the page where we add access to the Apis that your application needs.
   - Click the **Add a permission** button and then,
   - Ensure that the **Microsoft APIs** tab is selected.
   - In the *Commonly used Microsoft APIs* section, click on **Microsoft Graph**

>    - In the **Delegated permissions** section, select the **User.Read** in the list. Use the search box if necessary.

Reviewer 2: ```suggestion
   - In the **Delegated permissions** section, select **User.Read** in the list. Use the search box if necessary.

```

    Also, `MainPage.xaml.cs`, when building the application, ensure that the following line of code: ```.WithUseCorporateNetwork(true)```

> ### Step 4: Run the sample

Reviewer: ```suggestion
### Step 3: Run the sample

```

- we plan to remove this limitation in the future by integrating the Web Account Manager (WAM)

## Alternate approach to use WithDefaultRedirectURI()

> In the current sample, WithRedirectUri("https://login.microsoftonline.com/common/oauth2/nativeclient") method is used. To use WithDeaultRedirectURI(), please follow below steps:

Reviewer 2: ```suggestion
In the current sample, `WithRedirectUri("https://login.microsoftonline.com/common/oauth2/nativeclient")` method is used. To use `WithDefaultRedirectURI()`, please follow below steps:

```

> 1. In `MainPage.XAML.cs`, Update WithRedirectUri with WithDefaultRedirectUri as shown in below lines of code:

Reviewer 2: ```suggestion
1. In `MainPage.XAML.cs`, Update `WithRedirectUri` with `WithDefaultRedirectUri` as shown in below lines of code:

```

**Current Code**
```
PublicClientApp = PublicClientApplicationBuilder.Create(ClientId)
    .WithAuthority("https://login.microsoftonline.com/common")
    .WithUseCorporateNetwork(false)
    .WithRedirectUri("https://login.microsoftonline.com/common/oauth2/nativeclient")
    .WithLogging((level, message, containsPii) =>
     {
         Debug.WriteLine($"MSAL: {level} {message} ");
     }, LogLevel.Warning, enablePiiLogging: false, enableDefaultPlatformLogging: true)
    .Build();

```
**Updated Code**
```
PublicClientApp = PublicClientApplicationBuilder.Create(ClientId)
    .WithAuthority("https://login.microsoftonline.com/common")
    .WithUseCorporateNetwork(false)
    .WithDefaultRedirectUri()
    .WithLogging((level, message, containsPii) =>
     {
         Debug.WriteLine($"MSAL: {level} {message} ");
     }, LogLevel.Warning, enablePiiLogging: false, enableDefaultPlatformLogging: true)
    .Build();
```

> 2.	Discover the callback URI for your app by adding the following line and setting a breakpoint on:    

Reviewer 2: where should this line be added?

> var redirectURI = Windows.Security.Authentication.Web.WebAuthenticationBroker.GetCurrentApplicationCallbackUri();    

Reviewer 2: ```suggestion
```csharp
var redirectURI = Windows.Security.Authentication.Web.WebAuthenticationBroker.GetCurrentApplicationCallbackUri();    
```

```
